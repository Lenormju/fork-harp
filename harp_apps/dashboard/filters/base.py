from typing import Optional

from multidict import MultiDictProxy

from harp.typing.storage import Storage
from harp_apps.dashboard.filters.utils import flatten_facet_value


class AbstractFacet:
    name = "name"

    def __init__(self):
        self.meta = {}

    @property
    def values(self):
        raise NotImplementedError

    def get_filter(self, raw_data: list):
        raise NotImplementedError

    def filter(self, raw_data: list):
        raise NotImplementedError

    def get_filter_from_query(self, query: MultiDictProxy):
        raise NotImplementedError

    def filter_from_query(self, query: MultiDictProxy):
        raise NotImplementedError


class AbstractChoicesFacet(AbstractFacet):
    choices = set()

    @property
    def values(self):
        return [{"name": choice, "count": self.meta.get(choice, {}).get("count", None)} for choice in self.choices]

    def get_filter(self, raw_data: list):
        query_endpoints = self.choices.intersection(raw_data)
        return list(query_endpoints) if len(query_endpoints) else None

    def get_filter_from_query(self, query: MultiDictProxy):
        raw_data = self._choices_from_query(query)
        return self.get_filter(raw_data)

    def filter(self, raw_data: list):
        return {
            "values": self.values,
            "current": self.get_filter(raw_data),
        }

    def filter_from_query(self, query: MultiDictProxy):
        raw_data = self._choices_from_query(query)
        return self.filter(raw_data)

    def _choices_from_query(self, query: MultiDictProxy):
        return flatten_facet_value(query.getall(self.name, []))


class FacetWithStorage(AbstractChoicesFacet):
    def __init__(self, *, storage: Storage):
        super().__init__()
        self.storage = storage


class AbstractMinMaxFacet(AbstractFacet):
    min: float = 0.0
    max: float = 100.0

    def __init__(self):
        self.meta = {}

    @property
    def values(self):
        return {
            "min": self.meta.get("min", None),
            "max": self.meta.get("max", None),
        }

    def filter(self, min, max):
        return {
            "values": ["min", "max"],
            "current": self.get_filter(min, max),
        }

    def filter_from_query(self, query: MultiDictProxy):
        min, max = self._min_max_from_query(query)
        return self.filter(min, max)

    def get_filter(self, min: Optional[float], max: Optional[float]):
        return {"min": min, "max": max}

    def get_filter_from_query(self, query: MultiDictProxy):
        min, max = self._min_max_from_query(query)
        return self.get_filter(min, max)

    def _min_max_from_query(self, query: MultiDictProxy):
        return query.get(self.name + "min", None), query.get(self.name + "max", None)
