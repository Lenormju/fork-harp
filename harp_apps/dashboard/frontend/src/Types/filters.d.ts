export type FilterValue = string | number | boolean | undefined

export type MinMaxFilter = {
  min?: number
  max?: number
}

export type Filter = Array<FilterValue> | MinMaxFilter | undefined
export type ArrayFilter = Array<FilterValue> | undefined
export interface Filters extends Record<string, Filter> {}
