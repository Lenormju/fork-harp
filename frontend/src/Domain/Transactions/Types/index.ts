export { type Transaction } from "./Transaction"
export { type TransactionMessage } from "./TransactionMessage"
export { type TransactionRequest } from "./TransactionRequest"
export { type TransactionResponse } from "./TransactionResponse"
