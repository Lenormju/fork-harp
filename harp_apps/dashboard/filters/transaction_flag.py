from .base import AbstractChoicesFacet


class TransactionFlagFacet(AbstractChoicesFacet):
    name = "flag"
    choices = {"favorite"}
