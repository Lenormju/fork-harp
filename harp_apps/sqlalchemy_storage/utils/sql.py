from sqlalchemy import text


async def run_sql(engine, sql, *, autocommit=True):
    if isinstance(sql, str):
        sql = text(sql)
    async with engine.connect() as conn:
        result = await conn.execute(sql)
        if autocommit:
            await conn.commit()
    return result
