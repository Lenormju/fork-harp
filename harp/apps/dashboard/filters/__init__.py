from .transaction_endpoint import TransactionEndpointFacet
from .transaction_method import TransactionMethodFacet
from .transaction_status import TransactionStatusFacet
from .utils import flatten_facet_value

__all__ = [
    "TransactionMethodFacet",
    "TransactionStatusFacet",
    "TransactionEndpointFacet",
    "flatten_facet_value",
]
