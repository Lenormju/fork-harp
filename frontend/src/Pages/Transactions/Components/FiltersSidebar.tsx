import tw, { styled } from "twin.macro"

import { useTransactionsFiltersQuery } from "Domain/Transactions"
import { Filter, Filters } from "Types/filters"

import { Facet } from "./Facets"

const StyledSidebar = styled.aside`
  ${tw`max-w-full w-full divide-y divide-gray-100 overflow-hidden bg-white shadow-sm ring-1 ring-black ring-opacity-5`}
  ${tw`text-gray-900 sm:text-sm`}
`
const ratings = [
  { name: "A++" },
  { name: "A+" },
  { name: "A" },
  { name: "B" },
  { name: "C" },
  { name: "D" },
  { name: "E" },
  { name: "F" },
]

interface FiltersSidebarProps {
  filters: Filters
  setFilters: (filters: Filters) => unknown
}

export function FiltersSidebar({ filters, setFilters }: FiltersSidebarProps) {
  const filtersQuery = useTransactionsFiltersQuery()

  const _createSetFilterFor = (name: string) => (value: Filter) => {
    setFilters({ ...filters, [name]: value })
  }

  const setEndpointFilter = _createSetFilterFor("endpoint")
  const setMethodFilter = _createSetFilterFor("method")
  const setStatusFilter = _createSetFilterFor("status")

  return (
    <StyledSidebar>
      {/* TODO implement search
      <input className="h-12 w-full border-0 bg-transparent px-4 focus:ring-0" placeholder="Search..." />
     */}

      {filtersQuery.isSuccess && filtersQuery.data.endpoint ? (
        <Facet
          title="Endpoint"
          name="endpoint"
          type="checkboxes"
          values={filters["endpoint"]}
          setValues={setEndpointFilter}
          meta={filtersQuery.data.endpoint.values}
        />
      ) : null}

      {filtersQuery.isSuccess && filtersQuery.data.method ? (
        <Facet
          title="Request Method"
          name="method"
          type="checkboxes"
          values={filters["method"]}
          setValues={setMethodFilter}
          meta={filtersQuery.data.method.values}
        />
      ) : null}

      {filtersQuery.isSuccess && filtersQuery.data.status ? (
        <Facet
          title="Response Status"
          name="status"
          type="checkboxes"
          values={filters["status"]}
          setValues={setStatusFilter}
          meta={filtersQuery.data.status.values}
        />
      ) : null}

      {/* TODO implement */}
      {filtersQuery.isSuccess && filtersQuery.data.period ? (
        <Facet
          title="Period"
          name="period"
          meta={[{ name: "Last 24 hours" }, { name: "Last 7 days" }, { name: "Last 15 days" }]}
          type="radios"
        />
      ) : null}

      {/* TODO implement */}
      {filtersQuery.isSuccess && filtersQuery.data.apdex ? (
        <Facet title="Performance Index" name="tpdex" meta={ratings} type="checkboxes" defaultOpen={false} />
      ) : null}
    </StyledSidebar>
  )
}
