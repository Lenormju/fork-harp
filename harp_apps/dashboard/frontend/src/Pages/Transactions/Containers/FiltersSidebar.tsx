import { useCallback, useEffect } from "react"
import { useLocation, useNavigate } from "react-router-dom"

import { useTransactionsFiltersQuery } from "Domain/Transactions"
import { Filter, Filters } from "Types/filters"
import { Pane } from "ui/Components/Pane"

import { Facet } from "../Components/Facets"

const ratings = [
  { name: "A++" },
  { name: "A+" },
  { name: "A" },
  { name: "B" },
  { name: "C" },
  { name: "D" },
  { name: "E" },
  { name: "F" },
]

interface FiltersSidebarProps {
  filters: Filters
  setFilters: (filters: Filters) => unknown
}

export function FiltersSidebar({ filters, setFilters }: FiltersSidebarProps) {
  const location = useLocation()
  const navigate = useNavigate()

  const filtersQuery = useTransactionsFiltersQuery()

  const _createSetFilterFor = (name: string) =>
    useCallback(
      (value: Filter) => {
        setFilters({ ...filters, [name]: value })
      },
      [name],
    )

  const setEndpointFilter = _createSetFilterFor("endpoint")
  const setMethodFilter = _createSetFilterFor("method")
  const setStatusFilter = _createSetFilterFor("status")
  const setFlagsFilter = _createSetFilterFor("flag")

  // Set filters from query parameters
  useEffect(() => {
    const filtersMap = {
      endpoint: setEndpointFilter,
      method: setMethodFilter,
      status: setStatusFilter,
      flag: setFlagsFilter,
    }
    const queryParams = new URLSearchParams(location.search)
    const updateFilter = (name: string, setFunction: (value: Filter) => void) => {
      const values = queryParams.getAll(name)
      if (values.length) {
        setFunction(values)
      }
    }

    for (const [name, setFunction] of Object.entries(filtersMap)) {
      updateFilter(name, setFunction)
    }
  }, [location.search, setEndpointFilter, setMethodFilter, setStatusFilter, setFlagsFilter])

  // Update query parameters from filters
  useEffect(() => {
    const queryParams = new URLSearchParams(location.search)

    const updateQueryParams = (filterName: string) => {
      if (!filters[filterName] || !filters[filterName]?.length) {
        queryParams.delete(filterName)
      } else {
        // Delete existing query parameters for the filter
        queryParams.delete(filterName)

        // Append new query parameters for the filter
        filters[filterName]?.forEach((value) => queryParams.append(filterName, String(value)))
      }
    }

    // Update query parameters for each filter
    Object.keys(filters).forEach(updateQueryParams)

    navigate({
      pathname: location.pathname,
      search: queryParams.toString(),
    })
  }, [filters, location.pathname, location.search, navigate])

  return (
    <Pane hasDefaultPadding={false} className="divide-y divide-gray-100 overflow-hidden text-gray-900 sm:text-sm">
      {/* TODO implement search
      <input className="h-12 w-full border-0 bg-transparent px-4 focus:ring-0" placeholder="Search..." />
     */}

      {filtersQuery.isSuccess && filtersQuery.data.endpoint ? (
        <Facet
          title="Endpoint"
          name="endpoint"
          type="checkboxes"
          values={filters["endpoint"]}
          setValues={setEndpointFilter}
          meta={filtersQuery.data.endpoint.values}
        />
      ) : null}

      {filtersQuery.isSuccess && filtersQuery.data.method ? (
        <Facet
          title="Request Method"
          name="method"
          type="checkboxes"
          values={filters["method"]}
          setValues={setMethodFilter}
          meta={filtersQuery.data.method.values}
        />
      ) : null}

      {filtersQuery.isSuccess && filtersQuery.data.status ? (
        <Facet
          title="Response Status"
          name="status"
          type="checkboxes"
          values={filters["status"]}
          setValues={setStatusFilter}
          meta={filtersQuery.data.status.values}
        />
      ) : null}

      {filtersQuery.isSuccess && filtersQuery.data.flag ? (
        <Facet
          title="Flags"
          name="flags"
          type="checkboxes"
          values={filters["flag"]}
          setValues={setFlagsFilter}
          meta={filtersQuery.data.flag.values}
        />
      ) : null}

      {/* TODO implement */}
      {filtersQuery.isSuccess && filtersQuery.data.period ? (
        <Facet
          title="Period"
          name="period"
          meta={[{ name: "Last 24 hours" }, { name: "Last 7 days" }, { name: "Last 15 days" }]}
          type="radios"
        />
      ) : null}

      {/* TODO implement */}
      {filtersQuery.isSuccess && filtersQuery.data.apdex ? (
        <Facet title="Performance Index" name="tpdex" meta={ratings} type="checkboxes" defaultOpen={false} />
      ) : null}
    </Pane>
  )
}
