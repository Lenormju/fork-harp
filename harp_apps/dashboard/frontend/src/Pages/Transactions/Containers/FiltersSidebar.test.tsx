import { describe } from "node:test"

import { expect, it } from "vitest"

import { Filters } from "Types/filters"
import { renderWithClient } from "tests/utils"

import { FiltersSidebar } from "./FiltersSidebar"
import { MemoryRouter } from "react-router-dom"

const filters: Filters = {}

void describe("FiltersSidebar", () => {
  it("renders the title and data when the query is successful", async () => {
    const result = renderWithClient(
      <MemoryRouter>
        <FiltersSidebar filters={filters} />
      </MemoryRouter>,
    )
    await result.findByText("Request Method")
    expect(result.container).toMatchSnapshot()
  })
})
