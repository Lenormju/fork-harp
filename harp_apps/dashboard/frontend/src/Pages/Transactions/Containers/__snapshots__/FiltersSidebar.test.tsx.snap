// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the title and data when the query is successful 1`] = `
<div>
  <div
    class="divide-y divide-gray-100 overflow-hidden text-gray-900 sm:text-sm css-o0j4j7"
  >
    <div
      class="px-4 py-3"
    >
      <fieldset
        name="endpoint"
      >
        <legend
          class="flex w-full cursor-pointer css-9ho516"
        >
          <span
            class="grow"
          >
            Endpoint
          </span>
          <svg
            aria-hidden="true"
            class="h-4 w-4 text-gray-600"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.5 15.75 7.5-7.5 7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </legend>
        <div
          class="mt-2 space-y-2 "
        >
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="endpoint1"
                name="endpoint1"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="endpoint1"
              >
                <div>
                  endpoint1
                  <span
                    class="text-xs text-gray-400"
                  >
                     (
                    10
                    )
                  </span>
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="endpoint2"
                name="endpoint2"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="endpoint2"
              >
                <div>
                  endpoint2
                  <span
                    class="text-xs text-gray-400"
                  >
                     (
                    20
                    )
                  </span>
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <div
      class="px-4 py-3"
    >
      <fieldset
        name="method"
      >
        <legend
          class="flex w-full cursor-pointer css-9ho516"
        >
          <span
            class="grow"
          >
            Request Method
          </span>
          <svg
            aria-hidden="true"
            class="h-4 w-4 text-gray-600"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.5 15.75 7.5-7.5 7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </legend>
        <div
          class="mt-2 space-y-2 "
        >
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="GET"
                name="GET"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="GET"
              >
                <div>
                  GET
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="POST"
                name="POST"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="POST"
              >
                <div>
                  POST
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="PUT"
                name="PUT"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="PUT"
              >
                <div>
                  PUT
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="DELETE"
                name="DELETE"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="DELETE"
              >
                <div>
                  DELETE
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="PATCH"
                name="PATCH"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="PATCH"
              >
                <div>
                  PATCH
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <div
      class="px-4 py-3"
    >
      <fieldset
        name="status"
      >
        <legend
          class="flex w-full cursor-pointer css-9ho516"
        >
          <span
            class="grow"
          >
            Response Status
          </span>
          <svg
            aria-hidden="true"
            class="h-4 w-4 text-gray-600"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.5 15.75 7.5-7.5 7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </legend>
        <div
          class="mt-2 space-y-2 "
        >
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="2xx"
                name="2xx"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="2xx"
              >
                <div>
                  2xx
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="3xx"
                name="3xx"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="3xx"
              >
                <div>
                  3xx
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="4xx"
                name="4xx"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="4xx"
              >
                <div>
                  4xx
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="5xx"
                name="5xx"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="5xx"
              >
                <div>
                  5xx
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <div
      class="px-4 py-3"
    >
      <fieldset
        name="flags"
      >
        <legend
          class="flex w-full cursor-pointer css-9ho516"
        >
          <span
            class="grow"
          >
            Flags
          </span>
          <svg
            aria-hidden="true"
            class="h-4 w-4 text-gray-600"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.5 15.75 7.5-7.5 7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </legend>
        <div
          class="mt-2 space-y-2 "
        >
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="flag1"
                name="flag1"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="flag1"
              >
                <div>
                  flag1
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
          <div
            class="relative flex gap-x-3 select-none cursor-pointer"
          >
            <div
              class="flex h-6 items-center"
            >
              <input
                checked=""
                class="h-4 w-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-600 cursor-pointer"
                id="flag2"
                name="flag2"
                type="checkbox"
              />
            </div>
            <div
              class="text-sm leading-6"
            >
              <label
                class="font-medium text-gray-900  cursor-pointer"
                for="flag2"
              >
                <div>
                  flag2
                  <button
                    class="text-gray-400 mx-1 font-medium text-xs"
                  >
                    only
                  </button>
                </div>
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <div
      class="px-4 py-3"
    >
      <fieldset
        name="performanceIndex"
      >
        <legend
          class="flex w-full cursor-pointer css-9ho516"
        >
          <span
            class="grow"
          >
            Performances Index
            <button
              class="text-gray-400 mx-1 font-medium text-xs"
            >
              reset
            </button>
          </span>
          <svg
            aria-hidden="true"
            class="h-4 w-4 text-gray-600"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m4.5 15.75 7.5-7.5 7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </legend>
        <div
          class="mt-2 space-y-2 "
        >
          <div
            class="css-1tvrs2d"
          >
            <div
              class="css-1bggt6d"
            >
              <input
                class="input css-9ckkig"
                max="80"
                min="0"
                step="10"
                type="range"
                value="0"
              />
              <input
                class="input css-9ckkig"
                max="80"
                min="0"
                step="10"
                type="range"
                value="80"
              />
            </div>
            <div
              class="css-1yxx0tz"
            >
              <div
                class="css-14pv6h1"
                style="left: 0%;"
              />
              <div
                class="css-6u3d9g"
              >
                <div
                  class="css-fhc53s"
                  style="left: 0%; right: 0%;"
                />
                <div
                  style="position: absolute; left: 0%; top: 50%;"
                >
                  <div
                    class="bg-teal-400 css-1mbp5y2"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    A++
                  </div>
                </div>
                <div
                  style="position: absolute; left: 12.5%; top: 50%;"
                >
                  <div
                    class="bg-emerald-400 css-8d6cle"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    A+
                  </div>
                </div>
                <div
                  style="position: absolute; left: 25%; top: 50%;"
                >
                  <div
                    class="bg-green-500 css-tsujdb"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    A
                  </div>
                </div>
                <div
                  style="position: absolute; left: 37.5%; top: 50%;"
                >
                  <div
                    class="bg-lime-500 css-1lvmgf3"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    B
                  </div>
                </div>
                <div
                  style="position: absolute; left: 50%; top: 50%;"
                >
                  <div
                    class="bg-yellow-500 css-1tqf926"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    C
                  </div>
                </div>
                <div
                  style="position: absolute; left: 62.5%; top: 50%;"
                >
                  <div
                    class="bg-amber-500 css-1dfhn4m"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    D
                  </div>
                </div>
                <div
                  style="position: absolute; left: 75%; top: 50%;"
                >
                  <div
                    class="bg-orange-500 css-12ueco"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    E
                  </div>
                </div>
                <div
                  style="position: absolute; left: 87.5%; top: 50%;"
                >
                  <div
                    class="bg-red-500 css-1ksflql"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    F
                  </div>
                </div>
                <div
                  style="position: absolute; left: 100%; top: 50%;"
                >
                  <div
                    class="bg-red-700 css-eb9nog"
                  />
                  <div
                    class="css-rx8plc"
                  >
                    G
                  </div>
                </div>
              </div>
              <div
                class="css-14pv6h1"
                style="left: 100%;"
              />
            </div>
          </div>
        </div>
      </fieldset>
    </div>
  </div>
</div>
`;
